### ISSUE TITLE
Many CUDA test failures

### ISSUE THREAD

**Issue by ibaned** [gahansen/Albany#182]
I'm going to separate this issue out from #179.

@ikalash says:
> I am finally back from travel and had a chance to look at the status of the Ride CUDA Albany CDash.  Unfortunately, most of the tests are failing now whereas before we had a clean dashboard: http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany&subproject=IKTRideAlbanyCUDA .  There are failed comparisons for some tests; others time out.  Anyone have any thoughts on what might be causing this?

I don't know, but `SteadyHeat1D_Tpetra` is failing, so something pretty fundamental is broken.

**Comment by ikalash** [comment]
@jewatkins and I just talked about this and we're not sure how to best look into the problem besides just using the method of bisection to figure out which commit broke the tests.  Since my nightly tests on Ride were not up last week, I'm not sure if the original 9/11 commit involving lambdas + Tpetra broke things, or something that came after.  I am pretty sure the issue is on the Trilinos side rather than the Albany side - there were no nontrivial commits on the Albany side b/w 9/11 and now.  Thoughts?

**Comment by ikalash** [comment]
@mhoemmen, any thoughts?

**Comment by mhoemmen** [comment]
@ikalash I'll take a look.  I've been working on thread-parallelizing Tpetra stuff.  Not sure why I bother with Tpetra these days but I guess I want to keep making progress.

**Comment by ikalash** [comment]
Thanks, @mhoemmen.  @jewatkins plans to try to figure out today which commit / range of commits caused the issue.  We'll keep you posted.

**Comment by tjfulle** [comment]
@ikalash, are the failed CUDA build/test logs archived somewhere?  I don't see the original failures looking at the Albany dashboard today.

**Comment by ikalash** [comment]
@tjfulle : last night's job timed out so the results did not get pushed.  I am going to run it manually right now and hopefully things will get pushed to CDash then.

**Comment by jewatkins** [comment]
@tjfulle The failed tests from Wednesday are here: http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany&date=2017-09-21&subproject=IKTRideAlbanyCUDA&filtercount=3&showfilters=1&filtercombine=and&field1=buildname/string&compare1=61&value1=(empty)&field2=site/string&compare2=61&value2=ride.sandia.gov&field3=buildstamp/string&compare3=61&value3=20170921-0500-Nightly&collapse=0

**Comment by mhoemmen** [comment]
Hm, I haven't been testing on ride regularly.  I do most of my CUDA development on a non-IBM workstation with a Kepler GPU.  I'm looking at the Wednesday test results and notice that _some_ linear solves converged, so this may not be something obvious.

**Comment by tjfulle** [comment]
I've done all of my pre-PR testing of Tpetra on ride, the last time I tested, all Tpetra+Sacado+MueLu tests passed.

**Comment by jewatkins** [comment]
The problem might have something to do with communication since `SteadyHeat2D_SERIAL_Tpetra` passes but `SteadyHeat2D_Tpetra` fails.

**Comment by mhoemmen** [comment]
@jewatkins Does it help to revert the following two Trilinos develop commits?

679c06adab2d3e1974078a2f34b1069571555a47
8c5a31c280b8c712b2e25770802259b56e859cd9

**Comment by jewatkins** [comment]
@mhoemmen I just tried commit 8c5a31c280b8c712b2e25770802259b56e859cd9 and `SteadyHeat2D_Tpetra` still seems to fail.

**Comment by ikalash** [comment]
@tjfulle : here are today's failures 

http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=59870

and passes 

http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlypassed&buildid=59870

**Comment by jewatkins** [comment]
The last time all tests passed on Ride was on Sept. 7th. I found this commit so far which passes: [c661b83a9e9f8369d1fcbf85928bc7335aaef2aa](https://github.com/trilinos/Trilinos/commit/c661b83a9e9f8369d1fcbf85928bc7335aaef2aa)

**Comment by jewatkins** [comment]
I tried commits [b095c93d65010ae9277f38dad2ddaa2291e6b1b3](https://github.com/trilinos/Trilinos/commit/b095c93d65010ae9277f38dad2ddaa2291e6b1b3) and [cabb1eca6186056e36e5cc8f70228c936d3ba883](https://github.com/trilinos/Trilinos/commit/cabb1eca6186056e36e5cc8f70228c936d3ba883) and they both gave errors when building Trilinos related to Tpetra_Details_unpackCrsMatrixAndCombine_MP_Vector_Cuda.cpp.

**Comment by mhoemmen** [comment]
@jewatkins Does Albany actually use Stokhos?

**Comment by ikalash** [comment]
@mhoemmen : there are some capabilities in Albany that use Stokhos, yes, but not with CUDA; so we could turn off Stokhos in the build on Ride if we wanted to.

**Comment by ikalash** [comment]
@mhoemmen  : per your comment, I'm able to get some intermediate builds by turning off Stokhos.  I'm working on identifying what commit / set of commits caused this issue.  Stay tuned...

**Comment by mhoemmen** [comment]
@ikalash Thanks!

**Comment by ikalash** [comment]
@mhoemmen : here is an update.

The following Trilinos works: 80110046ffaf6b583bfeab39007b42791d490bb9

This one from a few hours later does not: 1eb8ef76f432543e543caccd9a2d09e4316570a4 .  

When I take the second of these and revert the commits you mention above, Albany tests are still failing, so it does not fix the problem.  

I will continue trying to isolate which commit broke Albany.  Let me know if you have any thoughts / suggestions from the new information.

**Comment by mhoemmen** [comment]
Thanks @ikalash !  I will take a look at those commits tomorrow morning and try to figure out what's going on.  I appreciate that you're putting in all the time this weekend.  I had a couple other work projects this weekend so I wasn't able to join you earlier in fixing this.

**Comment by mhoemmen** [comment]
Hm, this worries me: https://github.com/kokkos/kokkos/issues/542

**Comment by ikalash** [comment]
@mhoemmen : I have narrowed it down further.

ce22a7edf97af24ca744897374e6d2c5106e7471 is broken
80110046ffaf6b583bfeab39007b42791d490bb9 is OK

This means one of these commits are what broke Albany with CUDA KokkosNode: 

ommit ce22a7edf97af24ca744897374e6d2c5106e7471
Merge: 8011004 edced10
Author: Mark Hoemmen <mhoemme@sandia.gov>
Date:   Mon Sep 11 22:14:51 2017 -0600

    Merge remote branch 'intermediate-repo/develop' into develop
    
    Build/Test Cases Summary
    Enabled Packages: TpetraCore
    Disabled Packages: PyTrilinos,Claps,TriKota
    Enabled all Forward Packages
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=1506,notpassed=0 (85.93 min)
    Other local commits for this build/test group: edced10, a4178df

commit edced10a6a2df0d78c6b6bc4886342eee54ff945
Author: Mark Hoemmen <mhoemme@sandia.gov>
Date:   Mon Sep 11 16:01:27 2017 -0600

    Tpetra::CrsMatrix: Implement "new" DistObject interface
    
    @trilinos/tpetra Make Tpetra::CrsMatrix implement the "new"
    Tpetra::DistObject interface.  This means two things.  First,
    CrsMatrix now provides nontrivial implementations of all methods in
    the "new" DistObject interface:
    
      - copyAndPermuteNew
      - packAndPrepareNew
      - unpackAndCombineNew
    
    Second, CrsMatrix::useNewInterface now returns true, which tells
    DistObject to use the "new" interface.
    
    CrsMatrix retains an implementation of the "old" DistObject interface
    for now, but that will eventually go away.

commit a4178df6dc6b3edfefc0775b466265d21375a5e8
Author: Mark Hoemmen <mhoemme@sandia.gov>
Date:   Mon Sep 11 15:01:44 2017 -0600

    Tpetra: Move castAwayConstDualView into Tpetra::Details
    
    @trilinos/tpetra Move castAwayConstDualView out of an anonymous
    namespace in Tpetra_MultiVector_def.hpp, into Tpetra::Details and into
    a new header file Tpetra_Details_castAwayConstDualView.hpp.  Other
    Tpetra files may now reuse this function.  It remains an
    implementation detail of Tpetra.

As @jewatkins mentioned above, it appears only parallel tests are failing for the simple SteadyHeat2D problem, which might be a hint as to the source.  I will wait to hear from @mhoemmen regarding how to proceed.  I'm happy to try other builds and/or possible fixes.

**Comment by mhoemmen** [comment]
Hm, I'm wondering if it could be the changes made to Ifpack2.  Does this commit work?  ce22a7edf97af24ca744897374e6d2c5106e7471

**Comment by mhoemmen** [comment]
I just remembered that Albany might actually make use of `Ifpack2::BlockRelaxation` for ice sheet simulations.  If so, you might want to check out these commits:

3380bb19cd957bc5cb4fbb1655571b23159f56a6
1eb8ef76f432543e543caccd9a2d09e4316570a4

**Comment by ikalash** [comment]
@mhoemmen  : I can try it once Ride comes back up.  The errors show up for very basic Steady Heat problems that don't use anything special from Ifpack2 like the ice sheet simulations.

**Comment by ikalash** [comment]
@mhoemmen : actually I guess I tried already 

ce22a7edf97af24ca744897374e6d2c5106e7471 

As I report above, the code is broken with those commits.  I believe it is one of 3 Tpetra commits that broke things (see above).

**Comment by mhoemmen** [comment]
@ikalash I'm getting a Stokhos test failure in a CUDA build, `MPI_RELEASE`.  I will investigate that.

In the meanwhile, if reverting all three of those commits works, then I'll just revert them forever.

**Comment by ikalash** [comment]
@mhoemmen : There _was_ a Stokhos build error that cropped up around the same time as this one.  I just turned off Stokhos in Albany on Ride, since we don't need it.  Once Ride comes back up, I will verify that reverting the 3 commits resolves the issue from the Albany side and comment here.  Thanks for your help with this!

**Comment by mhoemmen** [comment]
Thanks @ikalash !

**Comment by mhoemmen** [comment]
Hm, `MPI_RELEASE` didn't tell me what was going on with that Stokhos test, so I've been rebuilding `MPI_DEBUG` for the past three hours.  It's still not done :( .

**Comment by mhoemmen** [comment]
It's done building now, and the test still fails, but it doesn't give me a useful stack trace.  Instead, it fails somewhere, and the trace happens in unrelated clean-up code (I've seen this before).

I'll keep at this.  @etphipp just said he would take a look at the Stokhos test too; that may be indirectly related, though Stokhos itself is innocent in this case :-) .

**Comment by mhoemmen** [comment]
I fixed a Tpetra issue (#1775), and @etphipp fixed something too (#1776).  The Stokhos test in question now passes in a CUDA build.  #1776 only relates to Sacado or Stokhos Scalar types, but #1775 was a legit CUDA bug relating to a missing `execution_space::fence()` between writing UVM memory on device, and reading it on host.  It could very well be that the commits in question that we were thinking about reverting were correct, but somehow manifested #1775.  At least, I'm hoping that :-) .

**Comment by ibaned** [comment]
@mhoemmen I updated your comment to link to Trilinos issues, extra syntax is needed since we're in the Albany repo here.

**Comment by mhoemmen** [comment]
Thanks @ibaned !

**Comment by ikalash** [comment]
@mhoemmen : Ride is back up so I can test potential fixes to Albany.  It looks like I cannot revert what I believe to be the problematic commits, as there are conflicts when I try.  Perhaps you can help me with this?

**Comment by mhoemmen** [comment]
We're working on this offline.

**Comment by ikalash** [comment]
@mhoemmen : any updates on this issue?  The Albany CUDA tests are still failing in the CDash.  I'd like to get them passing again soon so that if other issues are introduced that show up only with CUDA are caught by the test harness.

**Comment by mhoemmen** [comment]
Hi @ikalash !  No updates yet.  I was tempted just to roll back all those commits, but @kddevin preferred that we push through and figure out what's going on.  @tjfulle is working on a possibly related issue.  I'm not full time on Tpetra this FY, but I did add a test that should exercise Albany-like cases.  It passes for me on CUDA, but I'll build and run on ride as soon as I get the chance.

**Comment by ikalash** [comment]
Thanks for the update, @mhoemmen .  It would be nice if you could reproduce the issue on a smaller problem, so I'm interested what happens with your test on Ride.  If the issue shows up only in Albany, I'm happy to try to help one of you guys run the code to try to track it down.  

Any idea of a timeline for getting this resolved?  As I said above, I get worried when nightly tests are not running correctly for awhile b/c there is the chance that something else will break them and we will not catch it - leading to an even bigger nightmare in terms of tracking down / fixing things.

**Comment by mhoemmen** [comment]
@ikalash I'm not sure of the timeline yet, but we'll keep working on it.

**Comment by mhoemmen** [comment]
@ikalash We may have fixed this.  See discussion here:

https://github.com/trilinos/Trilinos/issues/1752

**Comment by ikalash** [comment]
@mhoemmen : that would be great if it did.  It looks like the nightly job on Ride last night still had failed tests (http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany&subproject=IKTRideAlbanyCUDA), but it was started at 2PM, which sounds like it was before you pushed the changes.  I will pull the code there, rebuild, rerun tests and see what happens.  Will keep you posted.

**Comment by tjfulle** [comment]
@ikalash, if you haven't already, I would hold off for just a little bit.  It seems the merge commit  [e385910](https://github.com/trilinos/Trilinos/commit/e385910a28504f745e102fd7c097788593f364c3) ended up reverting the changes that @mhoemmen was referring.  We're trying to figure out why and how to back out the commit (see comments at the bottom of the link).

@mhoemmen, could you weigh in over there on how you think it might be best to back out of the mess?

**Comment by mhoemmen** [comment]
@tjfulle I'm not sure if I have the authority to order people to revert their commits.  The easiest solution would be to restore the files themselves.  I'll comment there.

**Comment by mhoemmen** [comment]
@tjfulle From your latest comments, it looks like you're doing just that (restoring the files).  I won't distract you further :-) .

**Comment by tjfulle** [comment]
@ikalash, you should be good to go now.  Looking at the original (Albany) issue, if their is a timeout problem, the latest changes to Tpetra *should* fix that.  I don't know about the other problems, but we'll see!

**Comment by tjfulle** [comment]
@mhoemmen said:

> I'm not sure if I have the authority to order people to revert their commits

Yikes!  I hope it didn't come across that way!  I was hoping your git foo could help fix the problem.  Ultimately, `rsync` came to the rescue.

**Comment by mhoemmen** [comment]
@tjfulle wrote:
> I was hoping your git foo could help fix the problem.

I tried a `git revert` on the offending commit, and git said "NO" (in more words than that -- something about not wanting to revert a merge commit, which I found weird).  My proposed solution was going to be rolling back and just copying the file, which is what you did ;-) .

**Comment by ikalash** [comment]
@tjfulle , @mhoemmen : I have updated the codes on Ride and the tests are currently running.  They haven't completed yet, but I have looked at the output thus far, and unfortunately, a bunch are still failing :(.  Here is the Trilinos version I have: 0a3fe0d7f743d631d780bd760b80a0b3152a1ad4 .

**Comment by mhoemmen** [comment]
@ikalash Thanks for the update!  We are working a few issues right now.  The biggest one is Sierra / Trilinos integration; that's almost done.

**Comment by tjfulle** [comment]
@mhoemmen, I'll spend the next few days pounding on this.  I have a working sha and a broken sha.  I hope to find the source before my boys arrive on the 17th.

**Comment by ikalash** [comment]
Sounds great.  Thanks for working on this, @tjfulle !

**Comment by mhoemmen** [comment]
@tjfulle wrote:
> @mhoemmen, I'll spend the next few days pounding on this. I have a working sha and a broken sha. I hope to find the source before my boys arrive on the 17th.

If you have some time before then, perhaps today or tomorrow, we could do a knowledge transfer (or at least a patch transfer).  I'll send you an e-mail.  Thanks!

**Comment by tjfulle** [comment]
@ikalash and @mhoemmen (I'm sorry for the verbose notes, I want to make sure if my twins come earlier than planned that there is documentation)

I have found something, perhaps promising, by looking at the 1D heat transfer problem.  In this problem, during CrsMatrix sparse communication the first call to `Tpetra::Details::packCrsMatrix` packs a CrsMatrix whose rows appear to be have the following nonzero values: (-100, 200, -100).  The "fixed" and "broken" Trilinos' versions pack identical arrays (as we'd expect).  In the fixed version, (unsurprisingly) rows of (-100, 200, -100) are unpacked on the receiving end.  In the broken version, the unpacking follows identically to the fixed version *except* for the last row which is unpacked as (-100, 100, 0).  (I use the term row, but it's really just unpacking a pack buffer in to the crs array).

I don't yet know why the row is different: was it packed incorrectly?  is the unpack buffer not big enough, causing weirdness?  are the import/export indices wrong?  Any other ideas @mhoemmen?

Anyway, that's the update to now.  I'm looking in to the packing weirdness in more detail as we speak.

**Comment by tjfulle** [comment]
Adding to my last comment that I don't actually know that the rows of the CrsMatrix are of the form (-100, 200, -100), I was just guessing based on looking at massive dumps of data to the screen.  *But* regardless of the values, the last two values unpacked on the receiving end (in the broken version) are always wrong and the last value is always 0.

**Comment by ikalash** [comment]
Thanks for looking into it, @tjfulle !  This seems promising.  Does the incorrect unpacking happen only in parallel?  I remember a lot of tests were passing in serial but failing when run on multiple GPUs.  Your questions about packing/unpacking go beyond my knowledge of Tpetra; hopefully @mhoemmen has some ideas about what might be going on.

**Comment by tjfulle** [comment]
Packing/unpacking is for sparse parallel communication, so it would not be exercised in serial.

**Comment by ikalash** [comment]
Ah, yes, of course.  Sorry for the dumb question, it has been a long day... :(

**Comment by mhoemmen** [comment]
@tjfulle When you wrote "rows of (-100, 200, -100) are unpacked on the receiving end," does that refer to the values stored in the receive buffer, or to the final (combined) values in the target matrix?

It looks like the indices are getting messed up somehow.  Are the column Maps the same before and after?  That could also affect indexing.

**Comment by tjfulle** [comment]
@mhoemmen, you beat me to it, I was just about to write that these are the final values of the matrix - not the values being packed.  I'm investigating further, I'll add the column map to the list of things to check.

**Comment by tjfulle** [comment]
@mhoemmen, I think I might have more useful information.  The number of entries in each row to be unpacked is itself unpacked from the input buffer `imports`.  In the fixed version, this number is properly unpacked.  In the broken version, the unpacked value is 0.  The maximum number of entries to unpack in any one row is a reduction over all processes of the number of entries (the maximum) and is computed in `NumEntriesFunctor`, so this value is also 0.  Consequently, no values are actually unpacked.

So, it appears that the `imports` buffer to `Tpetra::Details::unpackAndCombine` is wrong.  I don't know where it has gone wrong though.  I don't actually know the code path to this point in the program.  I *think* the pack/unpack are being done as part of global assemble.

I'm off tomorrow, but will try to pick this up tomorrow night.  @mhoemmen, any thoughts?

**Comment by mhoemmen** [comment]
@tjfulle Do the packed values on the sending processes make sense?  I'm curious if we're not packing correctly, or if the problem is entirely on the unpack end.

**Comment by tjfulle** [comment]
@mhoemmen, I look at the packed values *before* packing and they make sense (i.e., are identical to the working program).  I have not verified that they are actually packed correctly.  I suppose I could do a quick manual unpack and check the value.

**Comment by mhoemmen** [comment]
@tjfulle Could it be that we're not counting sizeof the pack lengths, when packing them?  That would mess up unpack quite a bit.

**Comment by tjfulle** [comment]
@mhoemmen, I'm not sure?   Now that I have a smoking gun, I'll be able to answer more details.  I plan on looking at this more tonight.  But, the pack/unpack procedures are almost identical between the working/broken programs.  The biggest difference is the working program takes the "Old" code path and the broken takes the "New".  e.g., in the working program the code goes through `doTransferOld` and the broken through `doTransferNew` on their way to `do_unpack_and_combine_into_crs_matrix`.

**Comment by tjfulle** [comment]
@mhoemmen, one last thing before I take a break until later.  The test I am looking at passes if`false` is returned from `Tpetra::CrsMatrix::useNewInterface`.

**Comment by mhoemmen** [comment]
@tjfulle Are the packet counts getting lost somehow in various conversions between `Teuchos::Array*` and `Kokkos::*View`?

**Comment by tjfulle** [comment]
@mhoemmen, I don't think so. When I look at the inputs to the unpack routines, they all appear to be correct except the imports buffer.  It's unpacking from the imports buffer where things go wrong (on the receiving end)

**Comment by tjfulle** [comment]
@mhoemmen, I find something interesting/fishy.  When the number of entries in a row is unpacked from the imports buffer (in the broken version), it has a value of 0 (it should be 2).  But, in the packing procedure, rows with 0 values are never packed.  The number of entries unpacked from the imports should *never* be zero since we never pack zero rows.

**Comment by mhoemmen** [comment]
@tjfulle Did we never set the output argument in the PackTraits routine that unpacks the count?  Was it supposed to be passed out by reference but got passed by value instead?  I'll take a look... How hard would it be to isolate a test?

**Comment by tjfulle** [comment]
@mhoemmen, I'm not sure?  I don't think that is the case with the old interface.  I'm not positive about the new.  But, the new interface is used on other platforms without problems.  I believe the problem can be isolated to doTransferNew.  It appears that the imports buffer is good leaving packAndPrepareNew but is bad by the time it gets to unpackAndCombineNew. 

I can more easily walk you through what I've done on the phone and describe the problem that is breaking in detail. 
I have some build scripts that will build everything you need on `ride`.  Once the initial build is done (~1 hour for Trilinos+Albany), it only takes a minute or two rebuild between code modifications.  I'll send you the build scripts later (I'm away from a computer right now).  In fact, I can just open permissions to my directories on `ride` so you can just use what I have already built.  

I'll be around all weekend and all Monday.  Call/email me at your convenience.

**Comment by mhoemmen** [comment]
@tjfulle I wouldn't like to disturb your Sunday, but would Sunday (tomorrow) afternoon be a good time to chat?  Otherwise, I'll call Monday.  Thanks so much!

**Comment by tjfulle** [comment]
@mhoemmen, I replied via email.  Short version: Monday is looking crazy.  Why don't we plan on tomorrow at 2:45?

**Comment by mhoemmen** [comment]
I wonder if this has something to do with sync'ing of `numPacketsPerLID` and the `commOnHost` parameter in `Tpetra::DistObject`.  How do those routines decide whether to look at host or device data for unpacking?

**Comment by tjfulle** [comment]
@mhoemmen and I met and spoke about this issue this afternoon.  We believe we have narrowed down the problem to a small section of Tpetra.

**Comment by ikalash** [comment]
That's great @tjfulle !  Thank you for working on this over the weekend!

**Comment by rppawlo** [comment]
We are experiencing issues in CUDA/EMPIRE builds as well that look related to this.

**Comment by mhoemmen** [comment]
@rppawlo I'm working on it :-) Please feel free to share details over e-mail; thanks!

**Comment by ikalash** [comment]
Thank you for the update, @mhoemmen .  Let me know if you need any help with Albany once @tjfulle goes on paternity leave.  Also I'm happy to help test the fix in Albany when it's ready.

**Comment by mhoemmen** [comment]
@ikalash The test passes if I set the following environment variable:
```
$ export TPETRA_ASSUME_CUDA_AWARE_MPI=TRUE
```
This gives me an idea of what to test in Tpetra.  This may also help you as a temporary work-around, though I haven't tried this with the other tests yet.

**Comment by ibaned** [comment]
@mhoemmen does that macro force Tpetra to copy to the host before calling MPI, or to call MPI with GPU pointers? I would assume it makes Tpetra call MPI with GPU pointers.

**Comment by mhoemmen** [comment]
@ibaned It's an environment variable, not a macro :-) .  I've been moving towards more run-time control of Tpetra options.

The option, if `TRUE`, lets Tpetra assume that MPI can handle CUDA device buffers.  Tpetra only ever gives CudaSpace or HostSpace buffers to MPI, never CudaUVMSpace buffers.  If the option is `FALSE`, Tpetra will first copy to host before calling MPI.

**Comment by ibaned** [comment]
okay, so something is wrong with the "copy-to-host" codepath...

**Comment by mhoemmen** [comment]
I'm working on it :)

**Comment by mhoemmen** [comment]
Hm, `packCrsMatrixNew` ignores `commOnHost` (in `doTransferNew`) and always packs on device.  That's not a problem for the number-of-packets DualViews, since they always get sync'd to the appropriate space as a function of `commOnHost`.  However, `exports_` does _not_ get sync'd to host as a function of `commOnHost`.  That's no good.

Let me see if I can replicate this issue in Tpetra's tests, by setting the environment variable appropriately.

**Comment by mhoemmen** [comment]
Ugh, this takes forever... I'm simultaneously trying a possible fix in the Albany test framework that @tjfulle built up.

**Comment by mhoemmen** [comment]
Just setting the environment variable to `FALSE` does not make Tpetra's tests break.

**Comment by ikalash** [comment]
@mhoemmen : setting export TPETRA_ASSUME_CUDA_AWARE_MPI=TRUE does appear to fix the problem for steady heat!  I will run all the tests now to make sure everything works with that option.

**Comment by mhoemmen** [comment]
I think I get why Tpetra's "Albany182" test didn't catch this: the code path in question only gets exercised if the sparse matrix was built with a static graph.  Even if the matrix is fill complete, that code path wasn't getting called.  If this is true, then I should have a way to test this in Tpetra alone, and I shouldn't need ride to test it.

**Comment by ikalash** [comment]
Indeed with the environment variable export TPETRA_ASSUME_CUDA_AWARE_MPI=TRUE, all Albany tests pass on Ride with CUDA KokkosNode!

**Comment by ibaned** [comment]
@mhoemmen isn't there a CMake boolean equivalent of the `TPETRA_ASSUME_CUDA_AWARE_MPI` environment variable? If @ikalash is going to test with this on from now on (which I do think is a good idea), then I'd feel safer if its set at compile time.

**Comment by mhoemmen** [comment]
@ibaned wrote:
> [I]sn't there a CMake boolean equivalent of the TPETRA_ASSUME_CUDA_AWARE_MPI environment variable? If @ikalash is going to test with this on from now on (which I do think is a good idea), then I'd feel safer if its set at compile time.

Yes; you may set the CMake option `Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON` in order to force Tpetra to assume that your MPI implementation is CUDA aware.  You may still override this with the environment variable at run time, but it will default to your CMake setting.

It could be that `ride` is a cross-compiling situation, where the host node on which you build does not have a CUDA-aware MPI, but the compute nodes do.  That would affect the default value of the CMake option.  The CMake option exists in part for cross-compilation.

**Comment by ibaned** [comment]
Hmm I'm pretty sure I've confirmed with other Sandia codes that both of those GPU testbeds have CUDA-aware OpenMPI installations. I recommend @ikalash set that compile-time option `Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON`, since it is the more performant and more tested code path. Naturally we still want to fix the fallback code path, but Albany devs working on GPUs should get in the habit of setting this option.

**Comment by mhoemmen** [comment]
@ibaned That's fine for now.  Users who find that the MPI implementation is slow at accessing CUDA buffers can always control behavior via the environment variable.

**Comment by mhoemmen** [comment]
I modified the Tpetra "Albany182" test to exercise the static-graph case.  The test fails on CUDA and passes elsewhere for me, but it fails for a different reason (throws `std::logic_error`) in a debug build.  I will need to investigate the release build to see if it fails without throwing.

**Comment by tjfulle** [comment]
@mhoemmen said:

> I think I get why Tpetra's "Albany182" test didn't catch this: the code path in question only gets exercised if the sparse matrix was built with a static graph. Even if the matrix is fill complete, that code path wasn't getting called. If this is true, then I should have a way to test this in Tpetra alone, and I shouldn't need ride to test it.

Are the different code paths necessary once the matrix is fill complete?

**Comment by mhoemmen** [comment]
@tjfulle Actually no, the test should be "is the matrix fill complete?", not "does the matrix have a static graph."  I'm actually getting a different set of errors when I fix the Albany test, but I may have the test wrong; I'll need to think a bit more about it.

**Comment by mhoemmen** [comment]
@tjfulle I'm finding that `CrsMatrix::unpackRow` is encountering rows with zero entries.  Per our discussion yesterday, this should never happen.  I will dig into this more; it's probably related.

**Comment by mhoemmen** [comment]
@tjfulle FYI that only happens with CUDA.

**Comment by ikalash** [comment]
I'm fine closing the issue.  Thanks again, @mhoemmen  and @tjfulle for resolving this!

**Comment by rppawlo** [comment]
@mhoemmen - Tpetra still has a big in it. Using that flag is only a work around, right? Has a ticket been opened on the Trilinos side to track this since this ticket has been closed?

**Comment by mhoemmen** [comment]
@rppawlo : see https://github.com/trilinos/Trilinos/issues/1860

**Comment by mhoemmen** [comment]
I may have fixed this in my pushes to develop today.  I'll try it out in Albany as soon as I get the chance.

**Comment by ikalash** [comment]
@mhoemmen : cool, thanks!  Let me know what happens when you've had a chance to test it out on Albany.

**Comment by mhoemmen** [comment]
I'm having a wee bit of trouble redoing the build in @tjfulle 's directory, so I'll copy the scripts over and redo in my directory.  Tim was kind enough to set up all those scripts to make this easy to do.

It looks like today's Albany Dashboard for `ride` has link errors; it appears not to be finding MPI:

http://cdash.sandia.gov/CDash-2-3-0/viewBuildError.php?buildid=61339

**Comment by ikalash** [comment]
@mhoemmen : you are right, something in Trilinos broke Albany + CUDA again... sigh.  What's strange is despite the link error the Albany tests all still run.  Anyway, I have opened this as a separate issue (#195).
